### ISSUE TITLE
accessibility tags

### ISSUE THREAD

**Issue by jucasoft** [ag-grid/ag-grid#6150]
My company provided us with the new codes for the latest version of ag-grid; we went from version 27 to version 29 of ag-grid. A colleague of mine is using JAWS 2020 and can no longer see the data in the table. 

have you changed the ARIA accessibility tags?

**Comment by jucasoft** [comment]
the problem is that you have adopted the aria-describedby & aria-hidden="true" tag:
``` HTML
<div
  comp-id="395"
  aria-describedby="cell-395"
  class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height"
  aria-colindex="4"
  tabindex="-1"
  col-id="task"
  role="gridcell"
  style="left: 153px; width: 200px"
>
  <div class="ag-cell-wrapper" role="presentation" aria-hidden="true">
    <span class="ag-cell-value" role="presentation" id="cell-395">bla bla bla</span>
  </div>
</div>
```

i created a script to change all aria-hidden tags for this emergency period:

```js
  @HostListener('window:keyup', ['$event'])
  keyEvent(event: KeyboardEvent) {
    if (event.key === 'x' && event.altKey && event.composed) {
      this._ngZone.runOutsideAngular(() => {
        this._ngZone.run(() => {
          const demoClasses = document.querySelectorAll(
            '.ag-center-cols-clipper .ag-cell-wrapper'
          );
          demoClasses.forEach((element) => {
            element.ariaHidden = 'false';
          });
        });
      });
    }
  }
```

my colleague, in order to access the table information must press ALT+x, the script sets all the "aria-hidden="true"" tags to false.

updating JAWS 2020 will take weeks, you have planned a way to make this change backwards compatible.

**Comment by n0r8** [comment]
Moreover because of that getByRole in react-testing-library doesn't work anymore.

**Comment by kiril-matev** [comment]
Hi @jucasoft,

If you're an AG Grid Enterprise customer we recommend submitting issues in our support system, so they get processed more quickly.

In order to ensure any issues are processed quickly, please submit them following the AG Grid Github bug report template. In this case, you've not provided a live sample, a list of steps and actual/expected behavior. Quite importantly, you've not clarified what data JAWS isn't seeing in the table.

Can you please provide a test case showing the issue? We're keen to understand the specific scenario and the actual behavior you're getting - you can reference one of our documentation examples so you don't have to build one from scratch.

Looking forward to your response.

**Comment by kiril-matev** [comment]
Hi @jucasoft,

I built a test case for this:
1. Open JAWS 2020
2. Open
https://www.ag-grid.com/javascript-data-grid/row-height/#example-auto-row-height
3. Click the top-left cell
**Actual & Expected**: JAWS Announcement "Row 0"
4. Press RIGHT ARROW key
5. Listen to announcement
**Actual**: JAWS announcement "right parent autoheight"
**Expected**: JAWS announcement "lorem ipsum dolor sit amet...." (cell text value)

This behavior is due to an issue in JAWS 2020, which was fixed in JAWS 2023, which announces the value as expected.

We made a change to ARIA tags to resolve bug AG-7487 in v29.0 to avoid the double announcement of values in cells with auto-height, selection checkbox, row drag, etc. The DOM generated is correct and is correctly announced (without double announcement) by JAWS 2023.

Please upgrade to JAWS 2023 to resolve this issue.

The workaround you've built for this may prevent other accessibility-related announcements. I would suggest only using this temporarily until you upgrade to JAWS 2023.

Thank you for reporting this!

**Comment by kiril-matev** [comment]
Hello @n0r8,

Thanks for bringing this up and letting us respond definitively to this point in front of a wider audience.

The change in AG Grid v29.0 which now sets aria-hidden="true" is correct and helps resolve bug AG-7487 which was about multiple announcements of the same cell content:
https://ag-grid.com/changelog/?searchQuery=7487

You can still use the queryByRole / getByRole/ getAllByRole methods, but you need to set the option to show hidden elements as documented here:
https://pub.dev/documentation/react_testing_library/latest/rtl.dom.queries/getByRole.html
https://pub.dev/documentation/react_testing_library/latest/rtl.dom.queries/getAllByRole.html

This makes the function call as follows:
getAllByRole('checkbox', { hidden: true }) 
getByRole('checkbox', { hidden: true }) 
 
This allows the query to return items that are hidden for screen readers and should return the correct elements and fix this issue.

We've looked into this further and there's no other way to resolve the original issue without making the change. Please update your test code as shown above by setting hidden: true when calling the getByRole/getAllByRole methods.

Thank you for using AG Grid!

**Comment by n0r8** [comment]
> Hello @n0r8,
> 
> Thanks for bringing this up and letting us respond definitively to this point in front of a wider audience.
> 
> The change in AG Grid v29.0 which now sets aria-hidden="true" is correct and helps resolve bug AG-7487 which was about multiple announcements of the same cell content: https://ag-grid.com/changelog/?searchQuery=7487
> 
> You can still use the queryByRole / getByRole/ getAllByRole methods, but you need to set the option to show hidden elements as documented here: https://pub.dev/documentation/react_testing_library/latest/rtl.dom.queries/getByRole.html https://pub.dev/documentation/react_testing_library/latest/rtl.dom.queries/getAllByRole.html
> 
> This makes the function call as follows: getAllByRole('checkbox', { hidden: true }) getByRole('checkbox', { hidden: true })
> 
> This allows the query to return items that are hidden for screen readers and should return the correct elements and fix this issue.
> 
> We've looked into this further and there's no other way to resolve the original issue without making the change. Please update your test code as shown above by setting hidden: true when calling the getByRole/getAllByRole methods.
> 
> Thank you for using AG Grid!

Kiril, I hope you know that I tried and it doesn't work. For some reason `getAllByRole('checkbox', { hidden: true })` doesn't work in that case. Maybe because of some bug in the reac-testing-librarry, but it hurts anyways.

**Comment by kiril-matev** [comment]
Hi @n0r8,

Thanks for this response.

Please report this as a bug with **react-testing-library** - they need to support this as standard functionality.

Thank you for your time!

**Comment by jucasoft** [comment]
> 
I @kiril-matev 

Let me get this straight, you introduced this aria-describedby tag from 27.bla.bla which is supported by JAWS from version 2023.

so it's been 2 years since the ag-grid is not accessible for people using JAWS?

congratulations, great job.

could you tell me the advantage of using the aria-describedby tag.

if I remove aria-describedby and aria-hidden=true the table works perfectly and is accessible, could you explain to me why you introduced it? i can't understand it.

**Comment by ianldgs** [comment]
It's great that JAWS supports it now, but I don't think JAWS supporting it makes using `aria-describedby` with a hidden element a "standard functionality", as mentioned by @kiril-matev.

Take a look at this comment: https://github.com/w3c/accname/issues/57#issuecomment-529048546
> The only attributes that are permitted to reference hidden elements are aria-labelledby and aria-describedby. 

OK... but:
> Also, this is only true if the referenced element is explicitly hidden using CSS display:none or visibility:hidden, or the HTML5 hidden attribute.

**Comment by kiril-matev** [comment]
Hi @jucasoft,

Thanks for this response.

I think you misunderstood my earlier message.

Please note that we've introduced this change in AG Grid 29.0 which was released in January 2023. JAWS 2023 was released 3 months before this, in October 2022 and there was no gap in accessibility support for JAWS in any version of AG Grid.

AG Grid has been and is still accessible for users with JAWS 2020. This change only affects AG Grid versions 29.0 and later (which was released a month ago) which require upgrading to JAWS 2023.

Even with this change, if you use JAWS 2020 with AG Grid 29.0 and later, grid cells are announced correctly by JAWS 2020 and only cells which use auto-height / checkbox selection / row drag element are affected by this bug in JAWS 2020 and 2021 (resolved in JAWS 2023).

Please note that we do not make decisions on the implementation of accessibility support based on different screen readers, as they use different heuristics to process the DOM and announce its content. Instead, we are building our accessibility support implementation on the WCAG standard and rely on screen readers to support it. Where there are gaps in support of the WCAG standard in screen readers, we recommend submitting issues with the screen readers. 

In this particular case, JAWS have already addressed this issue and you need to upgrade to JAWS 2023.

As I mentioned earlier, the workaround you've built to remove these aria tags may resolve this particular issue in JAWS 2020 but may prevent other accessibility-related announcements. I would suggest only using this temporarily until you upgrade to JAWS 2023.

Thank you for giving us the opportunity to clarify these important points.

**Comment by jucasoft** [comment]
Hi @kiril-matev

can you tell me which version should I use without this attribute?

In your repository I found an issue opened 2 years ago about an error due to aria-describedby:

[https://github.com/ag-grid/ag-grid/issues/4675](https://github.com/ag-grid/ag-grid/issues/4675)

I found a commit from about 3 years ago where aria-describedby is rendered

[https://github.com/ag-grid/ag-grid/commit/a68e11cb40813babc4b4e38b3628b57bc54fae24](https://github.com/ag-grid/ag-grid/commit/a68e11cb40813babc4b4e38b3628b57bc54fae24)

I can't find precise information of when you introduced this attribute in your changeLog system:

[https://www.ag-grid.com/changelog/](https://www.ag-grid.com/changelog/)

We have decided to renew the ag-grid license, due to improvements in accessibility (read on your site).

after migrating our applications, 2 of my colleagues are not able to use the applications, They are not users, they are people who actively participate in the development, should I tell them that they can't work?

I noticed the problem because the person using JAWS works next to my desk, I wonder how many other people in the world have this problem.

good job.

**Comment by jucasoft** [comment]
Thanks @ianldgs, so the problem is not with the "aria-describedby" attribute, but with the fact that they used the "aria-hidden=true" attribute => [descriptio](https://github.com/ag-grid/ag-grid/issues/6150#issuecomment-1441648583)

Now that we have confirmed that this is incorrect, can we promote this issue to a bug?

Can anyone tell me how to change the function:

```
ag-grid/grid-community-modules/core/src/ts/utils/aria.ts

export function setAriaDescribedBy(element: HTMLElement, describedby: string | undefined): void {
    const key = 'describedby';
    if (describedby) {
        setAriaAttribute(element, key, describedby);
    } else {
        removeAriaAttribute(element, key);
    }
}

export function setAriaHidden(element: HTMLElement, hidden: boolean): void {
    if (hidden) {
        setAriaAttribute(element, 'hidden', true);
    } else {
        removeAriaAttribute(element, 'hidden');
    }
}
```

is there any way to change these functions by overriding the javascript prototype method?

Thanks

**Comment by kiril-matev** [comment]
Hello @jucasoft,

We understand that the fix we added to v29.0 to resolve issue AG-7487 (look up in our changelog) has required updating JAWS versions to 2023 and updating tests for many AG Grid developers. This is why we entered a feature request to update the fix for the original bug so it doesn't use the aria-hidden attribute. 

We've logged this as:
**AG-8085 [Accessibility] Investigate how to avoid duplicated announcements in React and VoiceOver for cells with auto-height/selection checkbox/drag and drop/ row drag without using aria-hidden attribute**

We hope to work on this soon.

See whether this item will be part of our next release by checking the NEXT RELEASE checkbox on the product pipeline page:
https://www.ag-grid.com/ag-grid-pipeline/

**Now that this issue is recorded in our backlog, we will soon close the GitHub issue because we will not update it when this bug is fixed in a future version of AG Grid. The best way to get an update on this item is to sign up for AG Grid new release notifications. Once a new AG Grid version is released, look for the item reference above (AG-XXXX) on the changelog page to see if it's resolved in that version:
https://www.ag-grid.com/ag-grid-changelog/**

Thanks again for bringing this up with us - we appreciate your time!

**Comment by jucasoft** [comment]
Thanks @kiril-matev  , I'll keep asking for updates on this post whenever my colleagues who can't use ag-grid ask me for updates.

**Comment by jucasoft** [comment]
@StephenCooper 
@kiril-matev 

how long does it take to release the new version?
I've seen in the source code, the problem seems to have been fixed for a couple of weeks, but I don't see any new versions released.

I have two colleagues who can't use our applications since I migrated ag-grid to version 29

**Comment by jucasoft** [comment]
> Hello @jucasoft,
> 
> The source code gets updated on a rolling basis but a release only happens once we've had a chance to add, document and test the new features we've planned for a particular release. The release of the next version is currently planned before the end of next week (before March 24th 2023).
> 
> As I mentioned before, you can upgrade to JAWS 2023 **or** use a version of your application that uses AG Grid 28.x to resolve this in the meantime.

Thank you for your answer.

We don't have the license for "JAWS 2023", we asked the company to buy it.

If I have to downgrade to ag-grid version 28, why did we pay for version 29?

**Comment by kiril-matev** [comment]
Hi @jucasoft,

I think there's a misunderstanding - I only recommended switching to an earlier version of your application which uses AG Grid v28.x as a temporary measure until this issue is addressed. 

This would allow your users with JAWS 2020 to continue using your application without any interruption until we ship the new release and you upgrade to it, rebuild, test and redeploy your application.

**Comment by jucasoft** [comment]
After two weeks of waiting, do you advise me to go back to 28? I see you fixed the problem in the source code, but you don't intend to release even a demo version?

**Comment by jucasoft** [comment]
My colleague says she can read the data in the table after upgrading to the latest version.
